### PR TITLE
first impl of MAT

### PR DESCRIPTION
**Pull request by rezabyt**
Integrating MAT into domain-based code. It changed the data loader so that we have access to sample indices in the update function of the MAT algorithm, allowing us to retrieve the assignment of each sample to each bias classifier.

### CODE REVIEW

File: domainbed/algorithms.py [hunk 1]
```diff
@@ -119,6 +120,141 @@ def update(self, minibatches, unlabeled=None):
...
+       
+        self.classifier_a, self.opt_a = self.create_bias_classifier(self.featurizer.n_outputs, num_classes, 
+                                                                    self.hparams['nonlinear_classifier'])
+        self.classifier_b, self.opt_b = self.create_bias_classifier(self.featurizer.n_outputs, num_classes, 
+                                                                    self.hparams['nonlinear_classifier'])
+
+        self.y_tr_dynamic = []
+        self.y_tr_true = []
+        for train_loader in train_loaders:
+            train_loader_ys = [y for _, y in train_loader.dataset.underlying_dataset]
```
Comment (mpezeshki) [inline domainbed/algorithms.py:145]: do the y's always have the same order independent of the seed?

File: domainbed/algorithms.py [hunk 2]
```diff
@@ -119,6 +120,141 @@ def update(self, minibatches, unlabeled=None):
...
+            self.y_tr_dynamic.append(torch.LongTensor(train_loader_ys))
+            self.y_tr_true.append(torch.LongTensor(train_loader_ys))
+
+        # randomly assigns each training example from all environments to one of the bias classifiers.
+        self.assigns = []
+        for i in range(len(self.y_tr_dynamic)):
+            assignments = torch.zeros(len(self.y_tr_dynamic[i]), 1).bernoulli_(0.5)
+            self.assigns.append(assignments)
```
Comment (mpezeshki) [inline domainbed/algorithms.py:152]: so if we have n environments, then len(self.assigns) is n. right?

File: domainbed/algorithms.py [hunk 3]
```diff
@@ -119,6 +120,141 @@ def update(self, minibatches, unlabeled=None):
...
+        
+        self.num_classes = num_classes
+
+    
+    def create_bias_classifier(self, in_features, out_features, nonlinear_classifier):
+        classifier = networks.Classifier(in_features, out_features, nonlinear_classifier)
+        classifier.weight.data != 0.0
+        classifier.bias.data != 0.0
```
Comment (mpezeshki) [inline domainbed/algorithms.py:163]: I think these two lines are not correct

Comment (mpezeshki) [reply domainbed/algorithms.py:163]: Also it appears that your assumptions is that nonlinear_classifier is False.

File: domainbed/algorithms.py [hunk 4]
```diff
@@ -119,6 +120,141 @@ def update(self, minibatches, unlabeled=None):
...
+                                                                    self.hparams['nonlinear_classifier'])
+        self.classifier_b, self.opt_b = self.create_bias_classifier(self.featurizer.n_outputs, num_classes, 
+                                                                    self.hparams['nonlinear_classifier'])
+
+        self.y_tr_dynamic = []
+        self.y_tr_true = []
+        for train_loader in train_loaders:
+            train_loader_ys = [y for _, y in train_loader.dataset.underlying_dataset]
+            self.y_tr_dynamic.append(torch.LongTensor(train_loader_ys))
+            self.y_tr_true.append(torch.LongTensor(train_loader_ys))
```
Comment (mpezeshki) [inline domainbed/algorithms.py]: what is y_tr_true used for?

File: domainbed/algorithms.py (hunk 1 above)
Comment (rezabyt) [reply domainbed/algorithms.py:145]: The order of y's is the same as what "_SplitDataset" returns samples. This dataset takes the keys and returns samples from the "underlying_dataset", I stored the same order of y's from "underlying_dataset". I changed "_SplitDataset" so that it returns the keys used to return samples and I use those indices to retrieve the "y_tr_dynamic" and "assignment" in the "update" method.

Regarding the order of "train_loaders": the order doesn't matter in this case. I am using the same order of "train_loaders" as how batches are created for the "update" method. Please look at line 216 of "train.py", where batches are created.

File: domainbed/algorithms.py (hunk 4 above)
Comment (rezabyt) [reply domainbed/algorithms.py]: It is redundant; I will remove it in the next push, since we already have access to true labels in each batch.

File: domainbed/algorithms.py (hunk 2 above)
Comment (rezabyt) [reply domainbed/algorithms.py:152]: Yes, it's correct.

File: domainbed/algorithms.py (hunk 3 above)
Comment (rezabyt) [reply domainbed/algorithms.py:163]: Regarding the first comment: If the classifier is linear, it is only a `torch.nn.Linear(in_features, out_features)`, so there is no specific name for the linear layer like "classifier.fc..."

Regarding the second comment, you are right. Do you think it's necessary to set the weight and bias to zero, or should I just remove these lines?

**Review by rezabyt — COMMENTED** [review]
I just replied to the comments. Please take a look.

File: domainbed/algorithms.py (hunk 1 above)
Comment (mpezeshki) [reply domainbed/algorithms.py:145]: great!

File: domainbed/algorithms.py (hunk 3 above)
Comment (mpezeshki) [reply domainbed/algorithms.py:163]: then I think it's better to check if the classifier is linear or not. For now, we can raise not implemented error if classifier is non-linear.
And yes, because of the flipping, it is necessary to set the weights and the biases to zero so at initialization there is zero chance for flipping.
